### PR TITLE
Lancehead TE proper support, second attempt

### PR DESCRIPTION
**Pull request by tildearrow**
Not yet implemented:
- Hardware profiles (and profile LED)

### CODE REVIEW

**Review by terricain — APPROVED** [review]
looks good to me. @z3ntu thoughts?

File: daemon/openrazer_daemon/hardware/mouse.py [hunk 1]
```diff
@@ -88,7 +88,32 @@ class RazerMambaTE(__RazerDeviceBrightnessSuspend):
...
+    USB_VID = 0x1532
+    USB_PID = 0x0060
+    HAS_MATRIX = True
+    MATRIX_DIMS = [1, 16]
+    METHODS = ['get_device_type_mouse', 'set_wave_effect',
+               'set_static_effect', 'set_spectrum_effect', 'set_reactive_effect', 'set_none_effect', 'set_breath_random_effect',
+               'set_breath_single_effect', 'set_breath_dual_effect', 'set_custom_effect', 'set_key_row', 'max_dpi',
+               'get_dpi_xy', 'set_dpi_xy', 'get_poll_rate', 'set_poll_rate']
 
     DEVICE_IMAGE = "https://assets.razerzone.com/eeimages/support/products/606/606_mambate_500x500.png"
```
Comment (z3ntu) [inline daemon/openrazer_daemon/hardware/mouse.py]: Please find an image for your device here

edit: you seem to have pasted the urls for this device down below. See the diff for details

File: driver/razermouse_driver.c [hunk 2]
```diff
@@ -732,6 +783,11 @@ static ssize_t razer_attr_read_set_brightness(struct device *dev, struct device_
         report.transaction_id.id = 0x3f;
         break;
 
+    case USB_DEVICE_ID_RAZER_LANCEHEAD_TE_WIRED:
+        // TODO fix
```
Comment (z3ntu) [inline driver/razermouse_driver.c]: What about this?

**Review by z3ntu — CHANGES_REQUESTED** [review]
See comments

Please also run `./scripts/generate_appstream_file.sh` and commit the results

File: daemon/openrazer_daemon/hardware/mouse.py (hunk 1 above)
Comment (tildearrow) [reply daemon/openrazer_daemon/hardware/mouse.py]: I am unable to find proper side/perspective images...

Comment (tildearrow) [reply daemon/openrazer_daemon/hardware/mouse.py]: Wait, thanks for pointing out. Will fix.

File: driver/razermouse_driver.c (hunk 2 above)
Comment (tildearrow) [reply driver/razermouse_driver.c]: Oops, forgot to remove a ToDo line.

**Comment by tildearrow** [comment]
I never knew generating an AppStream file was needed...

**Comment by tildearrow** [comment]
@z3ntu? I've pushed the changes you requested already.

**Review by terricain — APPROVED** [review]

**Comment by z3ntu** [comment]
Can we close this in favor of #833 ?

**Comment by tildearrow** [comment]
I'd say not to, because to be honest:
- #833 is inconsistent with the rest of supported Razer devices. Basically the only other device that uses logo/scroll separate commands is the Naga Hex V2.
- The official Synapse software only supports setting an effect for the entire mouse anyway, and while having the ability to set effects for every region in the mouse may look advantageous, it's pretty much a hack since the used commands seem to be legacy... (and once again, the only other device that does this is the Naga Hex V2)
  - Plus eventual custom effect support may surpass this hack.
- #833 does not provide a way to set an effect for the whole mouse (yet).
- #833 requires patching every front-end to support the wave effect for the logo and scroll wheel, plus left/right sides.
- Also, if #833 gets accepted this means the Lancehead TE would be the only device to support left/right side effects, which once again, is inconsistent...

**Comment by tildearrow** [comment]
Also, let me ask a question. Is there any reason for the preference of #833 other than it was released before this pull request?

**Comment by z3ntu** [comment]
Mostly because it actually implements support for the individual LED zones.

**Comment by tildearrow** [comment]
Where can I set an effect for the whole mouse, Synapse way then? And no, I am not going to take the effort of setting all the zones to the same effect.

Also, it looks more like a hack than a real solution (true custom effects)...
Furthermore, I repeat, that's inconsistent with the rest of devices. Please tell me which other device supports individual zone effects besides this and the Naga Hex V2.

I cannot forbid you from accepting his pull request, but please note that if you do I am **not** willing to work on getCurrentEffect and effect storage (#852) for the left and right sides.

**Comment by z3ntu** [comment]
> Where can I set an effect for the whole mouse, Synapse way then? And no, I am not going to take the effort of setting all the zones to the same effect.

That sounds like a feature request for the front ends, like a "synchronize all zones on the device" checkbox.

> Also, it looks more like a hack than a real solution (true custom effects)...

It isn't a hack. The device supports 4 separate hardware lighting effect zones, so that's what openrazer should support.

> Furthermore, I repeat, that's inconsistent with the rest of devices. Please tell me which other device supports individual zone effects besides this and the Naga Hex V2.

Imperator, DeathAdder Chroma, DeathAdder 2013, Naga Hex V2, Naga Chroma, Naga Hex, Naga Hex Red, Taipan, DeathAdder Elite, Naga 2014, Orochi 2011, Abyssus V2, DeathAdder 3500

**Comment by terricain** [comment]
If the device supports multiple led zones in hardware, we should support that

**Comment by tildearrow** [comment]
> That sounds like a feature request for the front ends, like a "synchronize all zones on the device" checkbox.

Wouldn't it be better to implement it in the driver? Less commands sent means less overhead.

> Imperator, DeathAdder Chroma, DeathAdder 2013, Naga Hex V2, Naga Chroma, Naga Hex, Naga Hex Red, Taipan, DeathAdder Elite, Naga 2014, Orochi 2011, Abyssus V2, DeathAdder 3500

With the exception of the DeathAdders most of these are legacy devices released prior to the Mamba Chroma (the first device to use the one-color-for-the-whole-mouse standard).

OK, so if nightsky's support gets merged I'll try to overhaul the Mamba and Mamba TE code.

**Comment by tildearrow** [comment]
Closing to push the night sky.

**Pull request closed without merging**